Remove PMCS processing of availability attributes

chamons opened the pull request:
- This PR is part of the "remove" PMCS workset, this one focusing on all availability attribute processing.
- As part of the PR involves touching a massive number of files, looking at each individual commit may be useful. In particular "Remove PMCS preprocessing of attributes"
- This PR is _IN WORK_. It commits, but let's not ship it yet.

monojenkins commented:
Build failure

spouliot commented on src/appkit.cs:
```diff
@@ -7991,7 +7991,7 @@ interface NSMenuItemCell {
 	}
 
-	[Availability (Introduced = Platform.Mac_10_0 | Platform.Mac_Arch32)]
+	[IntroducedAttribute (PlatformName.MacOSX, 10, 0, PlatformArchitecture.Arch32)] 
```
minor: we can drop the `Attribute` suffix

spouliot commented on src/WKWebKit/Defs.cs:
```diff
@@ -101,7 +106,7 @@ public enum WKDataDetectorTypes : nuint {
 	}
 
-	[iOS (10,0)][Mac (10,12, only64: true)]
+	[iOS (10,0)][Mac (10,12, onlyOn64: true)]
```
how did that compile ? :|

chamons replied:
PMCS "fixed" it before it hit a compiler.

spouliot replied:
:lol:

spouliot commented on src/qtkit.cs:
```diff
@@ -262,7 +262,8 @@ interface QTCaptureDevice : NSCoding {
 		
-		[Availability (Introduced = Platform.Mac_10_5 | Platform.Mac_Arch32)]
+		[IntroducedAttribute (PlatformName.MacOSX, 10, 5, PlatformArchitecture.Arch32)] 
+		[DeprecatedAttribute (PlatformName.MacOSX, 10, 9)] 
```
where does the deprecated comes from ?

spouliot commented on src/qtkit.cs:
```diff
@@ -614,17 +615,20 @@ interface QTFormatDescription {
 	interface QTMedia {
-		[Availability (Introduced = Platform.Mac_10_3 | Platform.Mac_Arch32, Deprecated = Platform.Mac_10_9)]
+		[IntroducedAttribute (PlatformName.MacOSX, 10, 3, PlatformArchitecture.Arch32)] 
+		[DeprecatedAttribute (PlatformName.MacOSX, 10, 5)] 
```
mismatch, deprecated was 10.9 and now is 10.5

chamons replied:
I copy pasted a few of these before writing up macros, and I apparently screwed a few up...

spouliot commented on src/scenekit.cs:
```diff
@@ -2398,7 +2398,8 @@ interface SCNRenderer : SCNSceneRenderer, SCNTechniqueSupport {
 		[Export ("render")]
-		[Availability (Deprecated = Platform.Mac_10_11 | Platform.iOS_9_0)]
+		[Availability (Deprecated = Platform.Mac_10_11)]
+		[Availability (Deprecated = Platform.iOS_9_0)]
```
maybe it would be best, when replacing, to use the latest attribute ? i.e.
`[Deprecated (PlatformName.iOS, 9, 0)]`

chamons replied:
For sure. It just involves more than 20 minutes and 5 vim macros to bang that out.

If you think it's worth it, I can hack this out (in a new PR).

spouliot replied:
yeah, let's split this out in a few different PR, starting the the old, named attributes
and we can split the work across the remaining people not yet on vacation :)

monojenkins commented:
Build failure

timrisi commented on src/generator-attribute-manager.cs:
```diff
@@ -83,8 +85,56 @@ static Type ConvertType (System.Type type, ICustomAttributeProvider provider)
+		switch (attribute.AttributeType.Name) {
+		case "SinceAttribute":
+		case "iOSAttribute":
+			return AttributeConversionManager.ConvertPlatformAttribute (attribute, 2 /* PlatformName.iOS */).Yield ();
```
Would it be better to cast these from the PlatformName enums instead of hard-coding the numbers, simply to make it harder to accidentally put in the wrong value if changes need to be made in the future? (And so one wouldn't have to look up what values they need to use from the enum if they have to make changes).

timrisi commented on src/generator-attribute-manager.cs:
```diff
@@ -269,3 +322,261 @@ static bool IsSubclassOf (Type base_class, Type derived_class)
+		return CreateMajorMinorAttribute (DeprecatedAttributeType, platform, major, minor, arch, message);
+	}
+
+	public static System.Attribute CreateUnavilableAttribute (PlatformName platformName, byte arch = 0xff, string message = "")
```
This is the original misspelled method for Sebastian's comment about misspelled Unavailable

timrisi commented on src/ObjCRuntime/PlatformAvailability2.cs:
```diff
@@ -205,4 +205,249 @@ public sealed class UnavailableAttribute : AvailabilityBaseAttribute
+		Mac_10_11   = 0x10000000000b0000,
+		Mac_10_11_3 = 0x10000000000b0300,
+		Mac_10_12   = 0x10000000000c0000,
+		Mac_10_13   = 0x10000000000d0000,
```
No tags needed for the other subversions (10.1, 10.2, 11.1, 11.2, anything for 10.12.x or 10.13.x)?

timrisi commented:
Also helps if I remember I'm reviewing the original commit when I make comments so I don't review on stuff that's already done 😇

chamons commented:
I rewrote the conversion code to simplify it and add cases for PlatformArch that will be introduced in https://github.com/xamarin/xamarin-macios/pull/3101.

I also reverted the 2nd and 3rd commit, since they will be covered by that, going first.

This is thus dont-merge until at least it lands and is resolved/tested.

chamons commented:
Still looking into if I can remove the using hacks but unsure it'll get better...

monojenkins commented:
Build failure

chamons commented:
As I reverted the conversion bits now covered in https://github.com/xamarin/xamarin-macios/pull/3101 it will fail to build until that lands. This is expected.

monojenkins commented:
Build failure

monojenkins commented:
Build failure

spouliot commented:
Last commit produced a few errors -> ```21:54:04 /Users/builder/jenkins/workspace/xamarin-macios-pr-builder/src/gamekit.cs(1163,79): error CS1016: Named attribute arguments must appear after the positional arguments
```

chamons commented:
@spouliot https://github.com/xamarin/xamarin-macios/pull/3096#issuecomment-352830355

chamons commented:
Depends on this PRs landing first:

https://github.com/xamarin/xamarin-macios/pull/3125
https://github.com/xamarin/xamarin-macios/pull/3123

Else it won't build.

Locally I was able to run the non-bcl mac tests successfully with those applied.

monojenkins commented:
Build failure

chamons commented on src/generator-attribute-manager.cs:
```diff
@@ -170,12 +220,15 @@ static System.Attribute CreateAttributeInstance (CustomAttributeData attribute,
-			if (list == null)
-				list = new List<T> ();
-			list.Add ((T) CreateAttributeInstance (attributes [i], provider));
+			foreach (var newAttribute in CreateAttributeInstance (attributes [i], provider)) {
```
Doesn't that shadow the variable on 172?

VincentDondain replied:
I'm just wondering why you're not using the variable that was already created here, that's all (:

chamons replied:
I'm taking that existing attribute, feeding it into CreateAttributeInstance which may convert it into one or more new attributes and adding those.

Shadowing the existing variable seems strange. I could remove attrib, have the two locations do attributes[i] and use it I guess...

chamons commented on src/generator-attribute-manager.cs:
```diff
@@ -269,3 +322,261 @@ static bool IsSubclassOf (Type base_class, Type derived_class)
+	{
+		var ctor = attribType.GetConstructor (ctorTypes);
+		if (ctor == null)
+			throw ErrorHelper.CreateError (1058, "Internal error: could not find a constructor for the mock attribute '{0}'. Please file a bug report (https://bugzilla.xamarin.com) with a test case.", attribType.FullName);
```
I stole it from here:

https://github.com/xamarin/xamarin-macios/blob/master/src/generator-attribute-manager.cs#L133

and we have a bunch of these to update:

```
gg "file a bug report" | grep -v docs | wc -l
     131
```

So I'd rather someone did it in an unrelated PR.

chamons commented:
With both of the prereq PRs landing, this might actually build/run tests 🎉 finally.

monojenkins commented:
Build failure

spouliot commented on tests/introspection/ApiTypoTest.cs:
```diff
@@ -323,6 +323,7 @@ protected ApiTypoTest ()
 			"Ntsc",
 			"nuint",
 			"Ndef",
+			"Noi", // From NoiOSAttribute
```
hmm... that was not caught before because it was turned into a `[Unavailable]` right ?

chamons replied:
Correct. But I added [NoiOS] and friends to match the [Mac] and friends that derive from the right attribute (since we're having to change tooling, let's make all of the common shortcuts work).

chamons replied:
https://github.com/xamarin/xamarin-macios/pull/3096/files#diff-8f0ef92138e06cd064a011ea8aad8d8bR279

monojenkins commented:
Build failure

monojenkins commented:
Build failure

spouliot commented:
The xtro failures is because the PR adds (non public) types that have no namespaces (it's a required hack). I'll fix this (in xtro) in a separate PR. There's nothing else (from xtro) blocking this PR (it can be considered green).

marek-safar commented on src/ObjCRuntime/PlatformAvailability2.cs:
```diff
@@ -205,4 +205,83 @@ public sealed class UnavailableAttribute : AvailabilityBaseAttribute
+		{
+		}
+
+		public TVAttribute (byte major, byte minor, bool onlyOn64 = false)
```
What's the point of default onlyOn64 parameter `false` value ? I cannot think of case where it will be used

chamons replied:
I just copy pasted the existing API from PlatformAvailability.cs to keep the API. I could have searched for usage and removed some but:

- I'd also have to search all branches in flight, such as any release/xcode not merged to master
- You lose some symmetry if you can do [Mac (10, 12, onlyOn64 : true) but not [TV (11, 0, onlyOn64 : true)]

marek-safar replied:
My comment is not about API but about default false value for onlyOn64, why is it there?

chamons replied:
I'm not sure what you are asking. 

Technically we are matching the original platform availability API found here:

https://github.com/xamarin/xamarin-macios/blob/master/src/ObjCRuntime/PlatformAvailability.cs#L593

In general, we have that override to allow us to mark certain APIs as 64-bit only as Apple often loves to either remove 32-bit implementations or never provide them.

If you are asking why we have it = false when that gives it the same behavior as above we want:

[TV (11, 0)] to not be 64-bit only and if it was = true we'd get that. The common pattern is [TV (11, 0, onlyOn64 : true)] to make it more clear in bindings what the true is for.

Looking at it, we maybe could have collapsed it to one ctor with more default parameters like

(byte major, byte minor, byte point = 0, bool onlyOn64 = false)

But I was looking to match the old API as close as possible.

Also, the reason that we never needed this was that these were processed by PMCS and converted into other attributes before the compiler ever saw them (see deleted code).

marek-safar replied:
> If you are asking why we have it = false when that gives it the same behavior as above we want:

Yes as `[TV (11, 0)]` won't use this overload, the only way to use it is to use explicit parameter name but then the default value won't be used

rolfbjarne replied:
tvOS is always 64-bits, so the `onlyOn64` parameter is redundant for tvOS.

The opposite logic goes for watchOS (it's always 32-bits), but I guess that can change in the future, so the parameter is fine.
